packages/foam-vscode/src/features/link-completion.ts

```diff
@@ -33,7 +33,7 @@ export class CompletionProvider
       .text.substr(0, position.character);
 
     // eslint-disable-next-line no-useless-escape
-    const requiresAutocomplete = cursorPrefix.match(/\[\[([^\[\]]*?)/);
+    const requiresAutocomplete = cursorPrefix.match(/\[\[[^\[\]]*(?!.*\]\])/);
```

Comment:
damn I hate regex... can you parse this out for me and tell me what it does?

Reply:
😸 I hate regex, too... And fix chinese tag issue(#567) is my first time to use RegExp in real code...

`\[\[[^\[\]]*(?!.*\]\])` can be divided into 3 parts:
1. `\[\[` means cursorPrefix should matches `[[`
2. `[^\[\]]*` means matches  every character except `[` and `]`
3. `(?!.*\]\])` means should not have `]]` after any characters(`.*`)

There are lots of site to test RegExp online like [String.prototype.match()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match) 
And this is the output
![TestOnLinePicture](https://cdn.jsdelivr.net/gh/RobinKingPubForToken/I/img/20210423213833.png)

---

I think `/\[\[[^\]]*(?!.*\]\])/` (only except `]` in part 2 ) is enough,
but the orign code except both `[` and `]` ,so I kept them, too. 😸

Reply:
Here is my online test code
```javascript
const para1 = '[[note';
const para2 = '[[note]] has been closed'
const para3 = '[[note]] closed but with another [[new-note'
const regex = /\[\[[^\]]*(?!.*\]\])/;

console.log(para1.match(regex));
console.log(para2.match(regex));
console.log(para3.match(regex));
```
And only except `]` code is like 
```suggestion
    const requiresAutocomplete = cursorPrefix.match(/\[\[[^\]]*(?!.*\]\])/);
```
This code also works fine...

Reply:
Here is my online test code
```javascript
const para1 = '[[note';
const para2 = '[[note]] has been closed'
const para3 = '[[note]] closed but with another [[new-note'
const regex = /\[\[[^\]]*(?!.*\]\])/;

console.log(para1.match(regex));
console.log(para2.match(regex));
console.log(para3.match(regex));
```
And only except `]` code is like 
```suggestion
    const requiresAutocomplete = cursorPrefix.match(/\[\[[^\]]*(?!.*\]\])/);
```
This code also works fine...

Reply:
can you please leave a short comment on the regex, either with this description (condensed) or with a link to this explanation - thanks!

Reply:
OK, and I modified the comment below for explanation, to prevent the above description from being collapsed.